Signal/src/ViewControllers/ViewControllerUtils.m

```diff
@@ -0,0 +1,73 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "ViewControllerUtils.h"
+#import "PhoneNumber.h"
+#import "StringUtil.h"
+#import <SignalServiceKit/PhoneNumberUtil.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark -
+
+@implementation ViewControllerUtils
+
++ (void)phoneNumberTextField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+                      countryCode:(NSString *)countryCode
+{
+
+    // Phone numbers takes many forms.
+    //
+    // * We only want to let the user enter decimal digits.
+    // * The user shouldn't have to enter hyphen, parentheses or whitespace;
+    //   the phone number should be formatted automatically.
+    // * The user should be able to copy and paste freely.
+    // * Invalid input should be simply ignored.
+    //
+    // We accomplish this by being permissive and trying to "take as much of the user
+    // input as possible".
+    //
+    // * Always accept deletes.
+    // * Ignore invalid input.
+    // * Take partial input if possible.
+
+    NSString *oldText = textField.text;
+    // Construct the new contents of the text field by:
+    // 1. Determining the "left" substring: the contents of the old text _before_ the deletion range.
+    //    Filtering will remove non-decimal digit characters like hyphen "-".
+    NSString *left = [oldText substringToIndex:range.location].digitsOnly;
+    // 2. Determining the "right" substring: the contents of the old text _after_ the deletion range.
+    NSString *right = [oldText substringFromIndex:range.location + range.length].digitsOnly;
+    // 3. Determining the "center" substring: the contents of the new insertion text.
+    NSString *center = insertionText.digitsOnly;
+    // 4. Construct the "raw" new text by concatenating left, center and right.
+    NSString *textAfterChange = [[left stringByAppendingString:center] stringByAppendingString:right];
+    // 4a. Ensure we don't exceed the maximum length for a e164 phone number,
+    //     15 digits, per: https://en.wikipedia.org/wiki/E.164
+    const int kMaxPhoneNumberLength = 15;
+    if (textAfterChange.length > kMaxPhoneNumberLength) {
+        textAfterChange = [textAfterChange substringToIndex:kMaxPhoneNumberLength];
```

Comment:
This is a change - the max length for the non-country calling code part of e164 is 15, so I'm starting to enforce that.  Previously users could type arbitrarily large numbers and it would break the layout.